File: server/api/scenes.py
"""Chapter scene-card API.

Scene cards are an additive planning layer.  Legacy ``Chapter.outline`` nodes
remain untouched; cards carry explicit chapter outline/body revisions instead.
"""

from datetime import datetime

from fastapi import APIRouter, Depends, HTTPException, Query, status
from pydantic import BaseModel, ConfigDict, Field
from sqlalchemy.ext.asyncio import AsyncSession

from api.auth import ProjectPermission, get_current_user, verify_chapter_assignment, verify_project_permission
from db.models_core import Chapter, User
from db.models_scene_cards import ChapterScene
from db.session import get_db
from domain.outlines import BodyPolicy
from services.scene_cards import (
    SceneBodyPolicyRequiredError,
    SceneChapterNotFoundError,
    SceneNotFoundError,
    SceneOrderError,
    ScenePlanRevisionConflictError,
    SceneReferenceError,
    SceneRevisionConflictError,
    archive_scene,
    create_scene,
    list_scenes,
    reorder_scenes,
    update_scene,
)

router = APIRouter()


class SceneFields(BaseModel):
    pov_entry_id: str | None = Field(default=None, max_length=32)
    location_entry_id: str | None = Field(default=None, max_length=32)
    goal: str = Field(default="", max_length=20_000)
    obstacle: str = Field(default="", max_length=20_000)
    turn: str = Field(default="", max_length=20_000)
    info_gain: str = Field(default="", max_length=20_000)
    emotion_shift: str = Field(default="", max_length=20_000)
    hook: str = Field(default="", max_length=20_000)
    status: str = Field(default="planning", pattern="^(planning|ready|written|needs_revision|archived)$")


class SceneCreateRequest(SceneFields):
    order: int = Field(default=1, ge=1)
    base_outline_rev: int = Field(ge=0)
    base_body_rev: int | None = Field(default=0, ge=0)
    body_policy: BodyPolicy | None = None


class SceneUpdateRequest(BaseModel):
    model_config = ConfigDict(extra="forbid")

    pov_entry_id: str | None = Field(default=None, max_length=32)
    location_entry_id: str | None = Field(default=None, max_length=32)
    goal: str | None = Field(default=None, max_length=20_000)
    obstacle: str | None = Field(default=None, max_length=20_000)
    turn: str | None = Field(default=None, max_length=20_000)
    info_gain: str | None = Field(default=None, max_length=20_000)
    emotion_shift: str | None = Field(default=None, max_length=20_000)
    hook: str | None = Field(default=None, max_length=20_000)
    status: str | None = Field(default=None, pattern="^(planning|ready|written|needs_revision|archived)$")
    expected_rev: int = Field(ge=1)
    base_outline_rev: int = Field(ge=0)
    base_body_rev: int | None = Field(default=0, ge=0)
    body_policy: BodyPolicy | None = None


class SceneReorderRequest(BaseModel):
    scene_ids: list[str] = Field(min_length=1, max_length=500)
    base_outline_rev: int = Field(ge=0)
    base_body_rev: int | None = Field(default=0, ge=0)
    body_policy: BodyPolicy | None = None


class SceneArchiveRequest(BaseModel):
    expected_rev: int = Field(ge=1)
    base_outline_rev: int = Field(ge=0)
    base_body_rev: int | None = Field(default=0, ge=0)
    body_policy: BodyPolicy | None = None


class SceneResponse(SceneFields):
    model_config = ConfigDict(from_attributes=True)

    id: str
    chapter_id: str
    order: int
    rev: int
    outline_rev: int
    body_rev: int | None
    archived_at: datetime | None
    created_at: datetime
    updated_at: datetime


def _response(scene: ChapterScene) -> SceneResponse:
    return SceneResponse.model_validate(scene)


async def _chapter_access(
    chapter_id: str,
    db: AsyncSession,
    user: User,
    permission: ProjectPermission,
) -> Chapter:
    try:
        chapter = await db.get(Chapter, chapter_id)
    except Exception:
        chapter = None
    if chapter is None or chapter.deleted_at is not None:
        raise HTTPException(status_code=404, detail={"code": "CHAPTER_NOT_FOUND"})
    await verify_project_permission(chapter.project_id, permission, user, db)
    if permission is ProjectPermission.MANAGE_OUTLINE:
        await verify_chapter_assignment(chapter, user, db)
    return chapter


def _raise_scene_error(error: Exception) -> None:
    if isinstance(error, SceneChapterNotFoundError | SceneNotFoundError):
        raise HTTPException(status_code=404, detail={"code": "SCENE_NOT_FOUND"}) from error
    if isinstance(error, SceneRevisionConflictError):
        raise HTTPException(
            status_code=status.HTTP_409_CONFLICT,
            detail={
                "code": "SCENE_REVISION_CONFLICT",
                "server_revision": error.actual,
                "client_revision": error.expected,
                "scene_id": error.scene.id,
            },
        ) from error
    if isinstance(error, ScenePlanRevisionConflictError):
        raise HTTPException(
            status_code=status.HTTP_409_CONFLICT,
            detail={
                "code": "SCENE_PLAN_REVISION_CONFLICT",
                "server_outline_revision": error.actual_outline,
                "client_outline_revision": error.expected_outline,
                "server_body_revision": error.actual_body,
                "client_body_revision": error.expected_body,
            },
        ) from error
    if isinstance(error, SceneBodyPolicyRequiredError):
        raise HTTPException(status_code=422, detail={"code": "BODY_POLICY_REQUIRED"}) from error
    if isinstance(error, SceneOrderError):
        raise HTTPException(status_code=409, detail={"code": "SCENE_ORDER_CONFLICT", "message": str(error)}) from error
    if isinstance(error, SceneReferenceError):
        raise HTTPException(status_code=422, detail={"code": "SCENE_REFERENCE_INVALID", "message": str(error)}) from error
    raise error


@router.get("/chapters/{chapter_id}/scenes", response_model=list[SceneResponse])
async def get_scenes(
    chapter_id: str,
    include_archived: bool = Query(default=False),
    user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
) -> list[SceneResponse]:
    await _chapter_access(chapter_id, db, user, ProjectPermission.VIEW)
    try:
        scenes = await list_scenes(db, chapter_id, include_archived=include_archived)
    except (SceneChapterNotFoundError, SceneNotFoundError) as error:
        _raise_scene_error(error)
    return [_response(scene) for scene in scenes]


@router.post("/chapters/{chapter_id}/scenes", response_model=SceneResponse, status_code=201)
async def post_scene(
    chapter_id: str,
    request: SceneCreateRequest,
    user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
) -> SceneResponse:
    chapter = await _chapter_access(chapter_id, db, user, ProjectPermission.MANAGE_OUTLINE)
    try:
        scene = await create_scene(
            db,
            chapter_id=chapter_id,
            order=request.order,
            fields=request.model_dump(exclude={"order", "base_outline_rev", "base_body_rev", "body_policy"}),
            base_outline_rev=request.base_outline_rev,
            base_body_rev=request.base_body_rev,
            body_policy=request.body_policy,
            project_id=chapter.project_id,
        )
        await db.commit()
        await db.refresh(scene)
    except Exception as error:
        await db.rollback()
        _raise_scene_error(error)
    return _response(scene)


@router.patch("/scenes/{scene_id}", response_model=SceneResponse)
async def patch_scene(
    scene_id: str,
    request: SceneUpdateRequest,
    user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
) -> SceneResponse:
    # Access is checked after loading the card in the service; first resolve its
    # chapter through a read-only query to avoid trusting a client project id.
    from sqlalchemy import select
    from db.models_scene_cards import ChapterScene

    existing = await db.scalar(select(ChapterScene).where(ChapterScene.id == scene_id))
    if existing is None:
        raise HTTPException(status_code=404, detail={"code": "SCENE_NOT_FOUND"})
    chapter = await _chapter_access(existing.chapter_id, db, user, ProjectPermission.MANAGE_OUTLINE)
    try:
        scene = await update_scene(
            db,
            scene_id_value=scene_id,
            expected_rev=request.expected_rev,
            fields=request.model_dump(exclude={"expected_rev", "base_outline_rev", "base_body_rev", "body_policy"}, exclude_unset=True),
            base_outline_rev=request.base_outline_rev,
            base_body_rev=request.base_body_rev,
            body_policy=request.body_policy,
        )
        await db.commit()
        await db.refresh(scene)
    except Exception as error:
        await db.rollback()
        _raise_scene_error(error)
    return _response(scene)


@router.post("/chapters/{chapter_id}/scenes/reorder", response_model=list[SceneResponse])
async def post_reorder(
    chapter_id: str,
    request: SceneReorderRequest,
    user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
) -> list[SceneResponse]:
    await _chapter_access(chapter_id, db, user, ProjectPermission.MANAGE_OUTLINE)
    try:
        scenes = await reorder_scenes(
            db,
            chapter_id=chapter_id,
            scene_ids=request.scene_ids,
            base_outline_rev=request.base_outline_rev,
            base_body_rev=request.base_body_rev,
            body_policy=request.body_policy,
        )
        await db.commit()
        for scene in scenes:
            await db.refresh(scene)
    except Exception as error:
        await db.rollback()
        _raise_scene_error(error)
    return [_response(scene) for scene in scenes]


@router.post("/scenes/{scene_id}/archive", response_model=SceneResponse)
async def post_archive(
    scene_id: str,
    request: SceneArchiveRequest,
    user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
) -> SceneResponse:
    from sqlalchemy import select
    from db.models_scene_cards import ChapterScene

    existing = await db.scalar(select(ChapterScene).where(ChapterScene.id == scene_id))
    if existing is None:
        raise HTTPException(status_code=404, detail={"code": "SCENE_NOT_FOUND"})
    chapter = await _chapter_access(existing.chapter_id, db, user, ProjectPermission.MANAGE_OUTLINE)
    try:
        scene = await archive_scene(
            db,
            scene_id_value=scene_id,
            expected_rev=request.expected_rev,
            base_outline_rev=request.base_outline_rev,
            base_body_rev=request.base_body_rev,
            body_policy=request.body_policy,
        )
        await db.commit()
        await db.refresh(scene)
    except Exception as error:
        await db.rollback()
        _raise_scene_error(error)
    return _response(scene)

File: app/src/api/scenes.ts
import { delay, request, USE_MOCK } from './http'

export type SceneStatus = 'planning' | 'ready' | 'written' | 'needs_revision' | 'archived'
export type SceneBodyPolicy = 'plan_only' | 'mark_body_for_revision'

export interface ChapterScene {
  id: string
  chapterId: string
  order: number
  povEntryId?: string
  locationEntryId?: string
  goal: string
  obstacle: string
  turn: string
  infoGain: string
  emotionShift: string
  hook: string
  status: SceneStatus
  rev: number
  outlineRev: number
  bodyRev?: number
  archivedAt?: string
  createdAt: string
  updatedAt: string
}

export type ScenePatch = Partial<Pick<ChapterScene, 'povEntryId' | 'locationEntryId' | 'goal' | 'obstacle' | 'turn' | 'infoGain' | 'emotionShift' | 'hook' | 'status'>>

interface SceneDto {
  id: string; chapter_id: string; order: number; pov_entry_id: string | null; location_entry_id: string | null
  goal: string; obstacle: string; turn: string; info_gain: string; emotion_shift: string; hook: string; status: SceneStatus
  rev: number; outline_rev: number; body_rev: number | null; archived_at: string | null; created_at: string; updated_at: string
}

function fromDto(row: SceneDto): ChapterScene {
  return {
    id: row.id, chapterId: row.chapter_id, order: row.order, povEntryId: row.pov_entry_id ?? undefined, locationEntryId: row.location_entry_id ?? undefined,
    goal: row.goal ?? '', obstacle: row.obstacle ?? '', turn: row.turn ?? '', infoGain: row.info_gain ?? '', emotionShift: row.emotion_shift ?? '', hook: row.hook ?? '',
    status: row.status, rev: row.rev, outlineRev: row.outline_rev, bodyRev: row.body_rev ?? undefined, archivedAt: row.archived_at ?? undefined,
    createdAt: row.created_at, updatedAt: row.updated_at
  }
}

const mockScenes = new Map<string, ChapterScene[]>()
function mockList(chapterId: string) { return mockScenes.get(chapterId) ?? [] }
function mockScene(chapterId: string, order: number, input: ScenePatch): ChapterScene {
  const now = new Date().toISOString()
  return { id: `scene-${chapterId}-${Date.now().toString(36)}`, chapterId, order, goal: '', obstacle: '', turn: '', infoGain: '', emotionShift: '', hook: '', status: 'planning', rev: 1, outlineRev: 0, createdAt: now, updatedAt: now, ...input }
}

export const scenesApi = {
  async list(chapterId: string): Promise<ChapterScene[]> {
    if (USE_MOCK) { await delay(60); return structuredClone(mockList(chapterId).sort((a, b) => a.order - b.order)) }
    return (await request<SceneDto[]>(`/chapters/${chapterId}/scenes`)).map(fromDto)
  },
  async create(chapterId: string, input: ScenePatch & { order?: number; baseOutlineRev: number; baseBodyRev?: number; bodyPolicy?: SceneBodyPolicy }): Promise<ChapterScene> {
    if (USE_MOCK) { await delay(80); const items = mockList(chapterId); const item = mockScene(chapterId, input.order ?? items.length + 1, input); items.push(item); mockScenes.set(chapterId, items); return structuredClone(item) }
    const { baseOutlineRev, baseBodyRev, bodyPolicy, order, ...fields } = input
    return fromDto(await request<SceneDto>(`/chapters/${chapterId}/scenes`, { method: 'POST', body: JSON.stringify({ ...toDto(fields), order: order ?? 1, base_outline_rev: baseOutlineRev, base_body_rev: baseBodyRev ?? 0, body_policy: bodyPolicy }) }))
  },
  async update(sceneId: string, patch: ScenePatch & { expectedRev: number; baseOutlineRev: number; baseBodyRev?: number; bodyPolicy?: SceneBodyPolicy }): Promise<ChapterScene> {
    if (USE_MOCK) {
      await delay(80); const item = [...mockScenes.values()].flat().find((scene) => scene.id === sceneId); if (!item) throw new Error('scene_not_found'); if (item.rev !== patch.expectedRev) throw new Error('scene_revision_conflict');
      Object.assign(item, patch); delete (item as unknown as Record<string, unknown>).expectedRev; delete (item as unknown as Record<string, unknown>).baseOutlineRev; delete (item as unknown as Record<string, unknown>).baseBodyRev; delete (item as unknown as Record<string, unknown>).bodyPolicy; item.rev += 1; item.updatedAt = new Date().toISOString(); return structuredClone(item)
    }
    const { expectedRev, baseOutlineRev, baseBodyRev, bodyPolicy, ...fields } = patch
    return fromDto(await request<SceneDto>(`/scenes/${sceneId}`, { method: 'PATCH', body: JSON.stringify({ ...toDto(fields), expected_rev: expectedRev, base_outline_rev: baseOutlineRev, base_body_rev: baseBodyRev ?? 0, body_policy: bodyPolicy }) }))
  },
  async reorder(chapterId: string, sceneIds: string[], baseOutlineRev: number, baseBodyRev?: number, bodyPolicy?: SceneBodyPolicy): Promise<ChapterScene[]> {
    if (USE_MOCK) { await delay(80); const items = mockList(chapterId); sceneIds.forEach((id, index) => { const item = items.find((scene) => scene.id === id); if (item) item.order = index + 1 }); return structuredClone(items.sort((a, b) => a.order - b.order)) }
    return (await request<SceneDto[]>(`/chapters/${chapterId}/scenes/reorder`, { method: 'POST', body: JSON.stringify({ scene_ids: sceneIds, base_outline_rev: baseOutlineRev, base_body_rev: baseBodyRev ?? 0, body_policy: bodyPolicy }) })).map(fromDto)
  },
  async archive(scene: ChapterScene, baseOutlineRev: number, baseBodyRev?: number, bodyPolicy?: SceneBodyPolicy): Promise<ChapterScene> {
    if (USE_MOCK) { await delay(60); scene.status = 'archived'; scene.archivedAt = new Date().toISOString(); scene.rev += 1; return structuredClone(scene) }
    return fromDto(await request<SceneDto>(`/scenes/${scene.id}/archive`, { method: 'POST', body: JSON.stringify({ expected_rev: scene.rev, base_outline_rev: baseOutlineRev, base_body_rev: baseBodyRev ?? 0, body_policy: bodyPolicy }) }))
  }
}

function toDto(fields: Record<string, unknown>): Record<string, unknown> {
  const map: Record<string, string> = { povEntryId: 'pov_entry_id', locationEntryId: 'location_entry_id', infoGain: 'info_gain', emotionShift: 'emotion_shift' }
  return Object.fromEntries(Object.entries(fields).map(([key, value]) => [map[key] ?? key, value]))
}
